// @aws-lambda-powertools/tracer/lib/config/EnvironmentVariablesService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvironmentVariablesService = void 0;
const _1 = require(".");
class EnvironmentVariablesService extends _1.ConfigService {
    constructor() {
        super(...arguments);
        // Environment variables
        this.awsExecutionEnv = 'AWS_EXECUTION_ENV';
        this.samLocalVariable = 'AWS_SAM_LOCAL';
        this.xRayTraceIdVariable = '_X_AMZN_TRACE_ID';
    }
    get(name) {
        return process.env[name]?.trim() || '';
    }
    getAwsExecutionEnv() {
        return this.get(this.awsExecutionEnv);
    }
    getCaptureHTTPsRequests() {
        return this.get(this.tracerCaptureHTTPsRequestsVariable);
    }
    getSamLocal() {
        return this.get(this.samLocalVariable);
    }
    getServiceName() {
        return this.get(this.serviceNameVariable);
    }
    getTracingCaptureError() {
        return this.get(this.tracerCaptureErrorVariable);
    }
    getTracingCaptureResponse() {
        return this.get(this.tracerCaptureResponseVariable);
    }
    getTracingEnabled() {
        return this.get(this.tracingEnabledVariable);
    }
    getXrayTraceId() {
        return this.get(this.xRayTraceIdVariable);
    }
}
exports.EnvironmentVariablesService = EnvironmentVariablesService;
