// @aws-lambda-powertools/logger/lib/config/ConfigService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigService = void 0;
/**
 * Abstract class ConfigService
 *
 * This class defines common methods and variables that can be set by the developer
 * in the runtime.
 *
 * @class
 * @abstract
 * @implements {ConfigServiceInterface}
 * @see https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime
 * @see https://awslabs.github.io/aws-lambda-powertools-typescript/latest/#environment-variables
 */
class ConfigService {
    constructor() {
        /**
         * @see https://awslabs.github.io/aws-lambda-powertools-typescript/latest/#environment-variables
         * @protected
         */
        this.currentEnvironmentVariable = 'ENVIRONMENT';
        this.logLevelVariable = 'LOG_LEVEL';
        this.sampleRateValueVariable = 'POWERTOOLS_LOGGER_SAMPLE_RATE';
        this.serviceNameVariable = 'POWERTOOLS_SERVICE_NAME';
    }
}
exports.ConfigService = ConfigService;
