// @aws-lambda-powertools/tracer/lib/middleware/middy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.captureLambdaHandler = void 0;
/**
 * A middy middleware automating capture of metadata and annotations on segments or subsegments ofr a Lambda Handler.
 *
 * Using this middleware on your handler function will automatically:
 * * handle the subsegment lifecycle
 * * add the `ColdStart` annotation
 * * add the function response as metadata
 * * add the function error as metadata (if any)
 *
 * @example
 * ```typescript
 * import { Tracer, captureLambdaHandler } from '@aws-lambda-powertools/tracer';
 * import middy from '@middy/core';
 *
 * const tracer = new Tracer({ serviceName: 'serverlessAirline' });
 *
 * export const handler = middy(async (_event: any, _context: any) => {
 *   ...
 * }).use(captureLambdaHandler(tracer));
 * ```
 *
 * @param target - The Tracer instance to use for tracing
 * @returns middleware object - The middy middleware object
 */
const captureLambdaHandler = (target) => {
    let lambdaSegment;
    const open = () => {
        lambdaSegment = target.getSegment();
        const handlerSegment = lambdaSegment.addNewSubsegment(`## ${process.env._HANDLER}`);
        target.setSegment(handlerSegment);
    };
    const close = () => {
        const subsegment = target.getSegment();
        subsegment?.close();
        target.setSegment(lambdaSegment);
    };
    const captureLambdaHandlerBefore = async () => {
        if (target.isTracingEnabled()) {
            open();
            target.annotateColdStart();
            target.addServiceNameAnnotation();
        }
    };
    const captureLambdaHandlerAfter = async (request) => {
        if (target.isTracingEnabled()) {
            target.addResponseAsMetadata(request.response, process.env._HANDLER);
            close();
        }
    };
    const captureLambdaHandlerError = async (request) => {
        if (target.isTracingEnabled()) {
            target.addErrorAsMetadata(request.error);
            close();
        }
    };
    return {
        before: captureLambdaHandlerBefore,
        after: captureLambdaHandlerAfter,
        onError: captureLambdaHandlerError
    };
};
exports.captureLambdaHandler = captureLambdaHandler;
