// @aws-lambda-powertools/metrics/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricResolution = exports.MetricUnit = exports.MAX_DIMENSION_COUNT = exports.MAX_METRIC_VALUES_SIZE = exports.MAX_METRICS_SIZE = exports.DEFAULT_NAMESPACE = exports.COLD_START_METRIC = void 0;
const COLD_START_METRIC = 'ColdStart';
exports.COLD_START_METRIC = COLD_START_METRIC;
const DEFAULT_NAMESPACE = 'default_namespace';
exports.DEFAULT_NAMESPACE = DEFAULT_NAMESPACE;
const MAX_METRICS_SIZE = 100;
exports.MAX_METRICS_SIZE = MAX_METRICS_SIZE;
const MAX_METRIC_VALUES_SIZE = 100;
exports.MAX_METRIC_VALUES_SIZE = MAX_METRIC_VALUES_SIZE;
const MAX_DIMENSION_COUNT = 29;
exports.MAX_DIMENSION_COUNT = MAX_DIMENSION_COUNT;
const MetricUnit = {
    Seconds: 'Seconds',
    Microseconds: 'Microseconds',
    Milliseconds: 'Milliseconds',
    Bytes: 'Bytes',
    Kilobytes: 'Kilobytes',
    Megabytes: 'Megabytes',
    Gigabytes: 'Gigabytes',
    Terabytes: 'Terabytes',
    Bits: 'Bits',
    Kilobits: 'Kilobits',
    Megabits: 'Megabits',
    Gigabits: 'Gigabits',
    Terabits: 'Terabits',
    Percent: 'Percent',
    Count: 'Count',
    BytesPerSecond: 'Bytes/Second',
    KilobytesPerSecond: 'Kilobytes/Second',
    MegabytesPerSecond: 'Megabytes/Second',
    GigabytesPerSecond: 'Gigabytes/Second',
    TerabytesPerSecond: 'Terabytes/Second',
    BitsPerSecond: 'Bits/Second',
    KilobitsPerSecond: 'Kilobits/Second',
    MegabitsPerSecond: 'Megabits/Second',
    GigabitsPerSecond: 'Gigabits/Second',
    TerabitsPerSecond: 'Terabits/Second',
    CountPerSecond: 'Count/Second',
};
exports.MetricUnit = MetricUnit;
const MetricResolution = {
    Standard: 60,
    High: 1,
};
exports.MetricResolution = MetricResolution;
