// @baselime/cdk/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Baselime = void 0;
var Baselime;
(function (Baselime) {
    let construct;
    let baselimeSecret;
    let serviceToken;
    let defaultChannel;
    let disableStackFilter;
    /**
     * Initialize Baselime CDK. Make sure to use this method in the beginning of the stack.
     *
     * @param {Construct} target
     * @param {BaselimeConfiguration} options
     * @example
     * import { Baselime } from '@baselime/cdk'
     * import * as cdk from 'aws-cdk-lib'
     * import { Construct } from 'constructs'
     *
     * export class ExamplesStack extends cdk.Stack {
     *   constructor(scope: Construct, id: string, props?: cdk.StackProps) {
     *     super(scope, id, props)
     *
     *     Baselime.init(this, {
     *       apiKey: process.env.BASELIME_API_KEY, // Ideally use SSM or Secrets Manager
     *		   defaultChannel: { type: "slack", targets: ["baselime-alerts"] },
     *     });
     */
    function init(target, options) {
        construct = target;
        baselimeSecret = options.apiKey;
        serviceToken = `arn:aws:lambda:${options.region || process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION || "eu-west-1"}:${options._account || "097948374213"}:function:baselime-orl-cloudformation`;
        defaultChannel = options.defaultChannel;
        disableStackFilter = options.disableStackFilter;
    }
    Baselime.init = init;
    function getConstruct() {
        return construct;
    }
    Baselime.getConstruct = getConstruct;
    function getApiKey() {
        return baselimeSecret;
    }
    Baselime.getApiKey = getApiKey;
    function getServiceToken() {
        return serviceToken;
    }
    Baselime.getServiceToken = getServiceToken;
    function getDefaultChannel() {
        return defaultChannel;
    }
    Baselime.getDefaultChannel = getDefaultChannel;
    function getDisableStackFilter() {
        return disableStackFilter;
    }
    Baselime.getDisableStackFilter = getDisableStackFilter;
})(Baselime = exports.Baselime || (exports.Baselime = {}));
